Assignment refactoring (#22)
DareData/lp-foundations: assignments/assignment_2/README.md

3. Assign a colleague to review your code.
4. After the review, merge the pull request.
5. Finally a pull request from `refactoring` to `main` or simply merge the `refactoring`.

# Assignment 2: Project Structure

We know that the things presented in the previous modules are orientative guidelines and real-life scenarios will slightly differ based on the project's specific needs. However, it is not uncommon to unnecessarily deviate too much from these.

## 1- Project restructuration

This exercise is very simple, we just want you to fish out some current or older project that you are or were part of and review its organization. What would you change? What would you keep? Why? 

## 2- Structure recognition

> In collecting a couple of examples for the project structure module we made a mess and lost track of the structure each of the projects. Give it a go and see if you can figure out which patterns do these projects follow.

Reviewer: Tend to agree with @MrCordeiro on having guided questions about the packages/repos. I like the idea of pointing to outside codebases as it promotes a skill that is often overlooked on most trainings (assessing the different nuances and preferences of other developers' coding style and choices). 

Adding to Fernando's questions, does it make sense to ask specific questions about specific modules? Anything that improves the ability to look into code that people didn't write themselves is a plus but let me know what you think.

[Link to mysterious project 1](https://github.com/qiuyu96/CoDeF/tree/main)

> [Link to mysterious project 2](https://github.com/RoberVega/mysterious_project_2)

Author: Nop, I accidentally made it private, now it is public.

This exercise is very simple, we just want you to fish out some current or older project that you are or were part of and review its organization. What would you change? What would you keep? Why? 

## 2- Structure analysis

> Again, a simple exercise. The idea this time is to practise your ability to navigate projects you have not written yourself. 

Reviewer 2: ```suggestion
Let's practice your ability to navigate projects you have not written yourself. 
```

It's a good practice to remove words like "simple", "trivial", etc while writing tutorials or documentation. It's very frustrating when we say something is simple when the user finds it to be anything but.

- How would you go about understanding the different components of a project for the first time? 
- What if you are looking for a specific feature, class, function, _etc._? How would you proceed then?

Enough chitchat, let us get our hands dirty! Your mission is to understand the structure of two of the most popular python libraries:
- [requests](https://github.com/psf/requests/tree/main),

> - [scikit-learn](https://github.com/scikit-learn/scikit-learn/tree/main). 

Reviewer 2: **Suggestion**: I would keep your 3rd project here as well: https://github.com/qiuyu96/CoDeF/tree/main

This example is interesting because it isn't a package, just a bunch of python scripts. Note how there are many entrypoints (one script for train, for hyperparameter optimization, etc) and one will have to read the documentation before being able to use it properly. 

Also note that this ties back to the `pyproject.toml` module: this repo was organized in a way that that all imports come from its sibling modules, so no functionality can be shared from parent modules to their children, only from the children to their parents. This was the first way we listed on how to solve the ModuleNotFound errors, and we don't recommend it: too much work for too little benefit.

Many repos coming out of academia are like this.

> By understanding we mean a very simple task: prepare a brief general description of the purpose of the library and some, also short, subdescriptions of each of the parts of the library.

Reviewer 2: **Suggestion**: make the tasks more explicit (while still keeping them open-ended)

```suggestion
By understanding we mean a very simple task: prepare a brief general description of the purpose of the library _just by looking at its folder structure_:

1. Can you identify the main folders? Without looking at the code (you can only look inside other folders) what is their purpose?
2.  Where is the main script of the package located? 
```
